qutip/solver/correlation.py

```diff
@@ -0,0 +1,520 @@
+        if ``c_ops`` are provided.
+    taulist : array_like
+        list of times for :math:`\tau`. taulist must be positive and contain
+        the element `0`.
+    c_ops : list
+        list of collapse operators, may be time-dependent for solver choice of
+        `me`.
+    a_op : Qobj
+        operator A.
+    args : dict
+        Dictionary of arguments to be passed to solver.
+    solver : str {'me', 'es'}
+        Choice of solver, `me` for master-equation, and `es` for exponential
+        series. `es` is equivalent to `me` with  ``options={"method": "diag"}``
+    options : dict, optional
+        Options for the solver. Only used with `me` solver.
+
+    Returns
+    -------
+    g2, G2 : tuple
+        The normalized and unnormalized second-order coherence function.
+
+    """
+    solver = _make_solver(H, c_ops, args, options, solver)
+
+    # first calculate the photon number
+    if state0 is None:
+        state0 = steadystate(H, c_ops)
+        n = np.array([expect(state0, a_op.dag() * a_op)])
+    else:
+        n = solver.run(state0, taulist, e_ops=[a_op.dag() * a_op]).expect[0]
+
+    # calculate the correlation function G2 and normalize with n to obtain g2
+    G1 = correlation_3op(solver, state0, [0], taulist,
+                         a_op.dag(), a_op.dag() * a_op, a_op)
+
+    g2 = G2 / (n[0] * n)
+    return g2, G2
+
+
+def _make_solver(H, c_ops, args, options, solver):
+    H = QobjEvo(H, args=args)
+    c_ops = [QobjEvo(c_op, args=args) for c_op in c_ops]
+    if solver == "me":
+        solver_instance = MeSolver(H, c_ops, options=options)
+    elif solver == "es":
+        options = {"method": "diag"}
+        solver_instance = MeSolver(H, c_ops, options=options)
+    elif solver == "mc":
+        raise ValueError("MC solver for correlation has been removed")
+    return solver_instance
+
+
+def correlation_3op(solver, state0, tlist, taulist, A=None, B=None, C=None):
+    r"""
+    Calculate the three-operator two-time correlation function:
+
+        :math:`\left<A(t)B(t+\tau)C(t)\right>`.
+
+    from a open system :class:`Solver`.
+
+    Note: it is not possible to calculate a physically meaningful correlation
+    where :math:`\tau<0`.
+
+    Parameters
+    ----------
+    solver : :class:`MeSolver`, :class:`BRSolver`, :class:`HEOMSolver`
+        Qutip solver for an open system.
+    state0 : Qobj
+        Initial state density matrix :math:`\rho(t_0)` or state vector
+        :math:`\psi(t_0)`.
+    tlist : array_like
+        List of times for :math:`t`. tlist must be positive and contain the
+        element `0`.
+    taulist : array_like
+        List of times for :math:`\tau`. taulist must be positive and contain
+        the element `0`.
+    A : Qobj, QobjEvo, optional, default=None
+        Operator A.
+    B : Qobj, QobjEvo, optional, default=None
+        Operator B.
+    C : Qobj, QobjEvo, optional, default=None
+        Operator C.
```

Comment:
I reworked it, merged all three into one entry to remove duplication.
I prefer the `None` to `1` since it shows better that you skip that operator, allowing to compute 2 operators correlation, instead of measuring a 3 operators correlation with one being the identity.